### PR TITLE
stop serving deleted APIs

### PR DESCRIPTION
**Pull request by deads2k**
This programatically enforces the removal of beta APIs that have expired.  Alpha levels have a grace period so that we can chase down their removal smoothly with CI by disabling the affected tests.  Controllers already had to handle the beta level detection, so they shouldn't be an issue.

/kind bug
/priority important-soon
@kubernetes/sig-api-machinery-bugs 

/hold the factorization is whacky, but I want to record the spot

```release-note
The kube-apiserver will no longer serve APIs that should have been deleted in GA non-alpha levels.  Alpha levels will continue to serve the removed APIs so that CI doesn't immediately break.
```

### CODE REVIEW

**Comment by deads2k** [comment]
/hold cancel

now with unit tests.

**Comment by sttts** [comment]
Since when do we fail on unused private methods?

pkg/controlplane/deleted_kinds_test.go:96:6: func storageNeverRemoved is unused (U1000)

File: pkg/controlplane/deleted_kinds.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	currentMajor int
+	currentMinor int
+	isAlpha      bool
+	// This is usually set for testing for which tests need to be removed.  This prevent insta-failing CI.
+	// Set KUBE_APISERVER_STRICT_REMOVED_API_HANDLING_IN_ALPHA to see what will be removed when we tag beta
+	strictRemovedHandlingInAlpha bool
+	// This is usually set by a cluster-admin looking for a short-term escape hatch after something bad happened.
+	// This should be made a flag before merge
+	// Set KUBE_APISERVER_SERVE_REMOVED_APIS_FOR_ONE_RELEASE to prevent removing APIs for one more release.
+	serveRemoveAPIsOneMoreRelease bool
```
Comment (sttts) [inline pkg/controlplane/deleted_kinds.go:41]: Removed

File: pkg/controlplane/deleted_kinds.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	versionsToRemove := sets.NewString()
+	for apiVersion, versionToResource := range versionedResourcesStorageMap {
+		resourcesToRemove := sets.NewString()
+		for resourceName, resourceServingInfo := range versionToResource {
+			if !e.shouldServe(resourceServingInfo) {
+				resourcesToRemove.Insert(resourceName)
+			}
+		}
+
+		for resourceName := range versionedResourcesStorageMap[apiVersion] {
```
Comment (sttts) [inline pkg/controlplane/deleted_kinds.go:135]: why this double loop, and `versionedResourcesStorageMap[apiVersion]` vs. `versionToResource`?

File: pkg/controlplane/deleted_kinds.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+
+	for _, apiVersion := range versionsToRemove.List() {
+		klog.V(1).Infof("Removing version %v.%v because it is time to stop serving it because it has no resources per APILifecycle.", apiVersion, groupName)
+		delete(versionedResourcesStorageMap, apiVersion)
+	}
+}
+
+func shouldRemoveResource(resourcesToRemove sets.String, resourceName string) bool {
```
Comment (sttts) [inline pkg/controlplane/deleted_kinds.go:155]: shouldRemoveResourceAndSubresources

File: pkg/controlplane/deleted_kinds.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	for _, resourceToRemove := range resourcesToRemove.List() {
+		if resourceName == resourceToRemove {
+			return true
+		}
+		// our API works on nesting, so you can have deployments, deployments/status, and deployments/scale.  Not all subresources
+		// server the parent type, but if the parent type (deployments in this case), has been removed, it's subresources should be removed too.
```
Comment (sttts) [inline pkg/controlplane/deleted_kinds.go:161]: serve the parent type

File: pkg/controlplane/instance.go [hunk 5]
```diff
@@ -25,6 +25,8 @@ import (
 	"strconv"
 	"time"
 
+	"k8s.io/component-base/version"
```
Comment (sttts) [inline pkg/controlplane/instance.go:28]: order

**Comment by sttts** [comment]
/lgtm
/approve
/hold

Just some nits. Feel free to unhold and retag after fixing.

File: pkg/controlplane/deleted_kinds.go (hunk 2 above)
Comment (deads2k) [reply pkg/controlplane/deleted_kinds.go:135]: > why this double loop, and `versionedResourcesStorageMap[apiVersion]` vs. `versionToResource`?

10 lines up, the godoc describes the args.  This mirrors the structure in `APIGroupInfo`, so per the comment `it's a map from version to resource to the storage.`.  If I change it here, the usage becomes more difficult and the results cannot be directly used.

**Comment by deads2k** [comment]
```
$ git fetch https://github.com/kubernetes/kubernetes.git pull/90983/head
From https://github.com/kubernetes/kubernetes
 * branch                    refs/pull/90983/head -> FETCH_HEAD
$ git merge --no-ff a6a6b17348909030b7563dd1c21fa166c668844b
merge: a6a6b17348909030b7563dd1c21fa166c668844b - not something we can merge
# Error: exit status 1
```

slow github repo update?

/retest

**Comment by liggitt** [comment]
replaced by https://github.com/kubernetes/kubernetes/pull/96525

/close

**Pull request closed without merging**